sdlib/d/gc/hpd.d

```diff
@@ -111,6 +111,55 @@ public:
 		return bestIndex;
 	}
 
+	uint shrinkAlloc(uint index, uint oldPages, uint delta) {
+		assert(delta < oldPages, "Invalid shrink pages count!");
+
+		release(index + oldPages - delta, delta);
+		allocCount++;
+
+		return oldPages - delta;
+	}
```

Comment:
Let me try to explain why we are doing many review cycles - hence things are not progressing fast, even through you are clearly able to come up with reasonable solution. It's the details, and these details can be gleaned by "listening" to the code. It is telling you things. Here are a few things that can be noticed reading this code:

1. The first thing to be noticed is that this does pretty much the same as release. It stand to reason that both should live one next to the other, so the code is easier to follow.
2. `shrink`'s API contains a lot of parameters that are not used by the HPD in any meaningful way. This tells me these computation do not belong in the HPD, but in shrink's caller.
3. `release` decrement `allocCount`, while this doesn't. So, instead of `shrink` calling `release`, and then canceling the decrement, it would be strictly better to have `release` call `shrink` and then decrement `allocCount`.

None of these remarks where predictable ahead of time, but they are obvious now by reading that piece of code.

This gets us to smaller PRs. When these details are lost in the middle of several hundred of line of code of changes, then do not stand up. When they are by themselves, they do.